README.md
# Portfolio Brusselaers Niko

this is a fullstack application with the function to receive noise level data from arduino sensors,
store all of this data inside a database, and display average noise levels over the course of an hour,day,week,month.
this was created to be used by universities, colleges, study cafe's, library, ... to setup this application to give
people an general overview how loud it is in certain room/spaces and to let plan to work in peace.
## Table of content
- [Portfolio Brusselaers Niko](#portfolio-brusselaers-niko)
  - [Table of content](#table-of-content)
  - [Installation](#installation)
    - [Requirements](#requirements)
    - [Setup](#setup)
  - [license](#license)
  - [links](#links)
## Installation
### Requirements
to install and run this application you will need the following:
  - docker: 
    - windows: https://docs.docker.com/desktop/install/windows-install/
    - linux: https://docs.docker.com/desktop/install/linux-install/
  - docker-compose:
    - windows(comes already installed with docker):https://docs.docker.com/desktop/install/windows-install/
    - linux: https://docs.docker.com/compose/install/linux/
  if you use arduino to record data:
  - arduino ide: https://www.arduino.cc/en/software

to make certain parts of this application available outside of your local network, you will need to open following port on your router:
  - Front-end website: 80
  - phpMyAdmin: 8080

if you want to you an arduino to collect data, use a arduino capable of connecting to wifi networks as for example:
  - Arduino Uno Wifi
  - Arduino ESP32
  - arduino ESP8266
when choosing a sound sensor, make sure the sensor is capable of detecting noise levels

### Setup
Create a new file ".env" at roof of file directory and add the following Parameters:
  - `MYSQL_USERNAME = "root"`
  - `MYSQL_ROOT_PASSWORD= "password for root"`
  - `MYSQL_DATABASE="database name"`

open shell terminal and navigate to the folder containing application files, and run the command `docker-compose up --build -d`

if you are using arduino devices to record data:
  - psuedo data: unzip psuedo data zip file and rename folder to `arduino_volume_sensor_mqtt`.
  open up `arduino_volume_sensor_mqtt.ino` inside folder.
  Update espMQTTClient and enter all with your wifi and broker credentials
  update room to the name of the room you are placing it in and set the desired device name in the deviceName variable.
  Connect your arduino to the computer and select the drivers the arduino needs and the usb port its been connected to.
  Upload your data to the arduino device.


  - sensor data: unzip psuedo data zip file and rename folder to `arduino_volume_sensor_mqtt`.
  open up `arduino_volume_sensor_mqtt.ino` inside folder.
  Update espMQTTClient and enter all with your wifi and broker credentials
  update room to the name of the room you are placing it in and set the desired device name in the deviceName variable,
  update the soundSensorPin variable to the pin where you going to connect the analog pin of the sound sensor.
  Connect your arduino to the computer and select the drivers the arduino needs and the usb port its been connected to.
  Upload your data to the arduino device. Connect a comptible sound sensor to the arduino

## license
  - [license](license.md)

## links
  - [contributing](CONTRIBUTING.md)
  - [Code of Conduct](CODE_OF_CONDUCT.md)
  - [changelog](changelog.md)
  - [references](references.md)

changelog.md
# changelog 

## [0.1.0] - 2022-12-06 

### added
 - MYSQL database
 - phpMyAdmin 
 - Mosquitto broker
 - MQTT api
   - inserting incomeing sensor data inside database
   - inserting incoming new rooms inside database

## [0.2.0] - 2022-12-08

### added
  - api
    - deleteRoom endpoint
    - getAllRooms endpoint
    - getRoomData endpoint

## [0.2.1] - 2022-12-16

### fixed

  - added a health check to see if the mysql database is fully opertional to prevent rest and mqtt api from crashing from being unable to       connect to database

## [0.3.1] - 2022-12-16

### added
  - api
    - addRoom endpoint

## [0.3.2] - 2022-12-16

### fixed
  - api
    - changed getRoomData from GET to POST endpoint
  - mysql
    - typo in sensorData table

### removed
  - api
    - receiveRoomData endpoint

## [0.4.2] - 2022-12-19

### added
  - api
    - login endpoint
    - loginWithId endpoint
  - mysql
    - users table with collumns: id, username and password

## [0.5.2] - 2022-12-20

### added
  - front-end website
    - add rooms to collect data from
    - remove rooms
    - view all noise level data from saved rooms

## [0.5.3] - 2022-12-20

### added
  - api
    - periodic restart to make sure the api stays online

## [0.5.4] - 2022-12-21

### changed
  - api
    - getAllRooms endpoint: refactor code to optimize

## [0.5.5] - 2022-12-21

### changed
  - api
    - getAllRooms endpoint: refactor code of endpoints to optimize:
      - getRoomData
      - addRoom
      - deleteRoom
      - login
      - loginWithId

## [0.5.6] 2022-12-21

  ### fixed
  - api
    - added error handeling for getRoomData

  ### removed
  - api
    - unnecessary console.log() from code 
    - unused updateRoomSensor endpoint

## [0.5.7] 2022-12-21

  ### updated
  - front-end
    - api url port to port forwarded port on the router

## [0.6.7] 2022-12-22

### added
  - front-end
    - testing for filterByDate function
  - arduino
    - psuedo data code to test out connection to broker and database is working
    - sensor data code to run to place devices in diffrent rooms and record data


images/front-end/src/math/getGraph.js
/**
 * object template
 * 
 * chartData{
 *      date(arr(str)),
 *      value(arr(int))
 * }
 */


/**
 * Read received data from room and create a chart.
 * 
 * 
 * @param chartData
 * @param duration(str) 
 * @returns
 */
const getGraph = (chartData, duration) => {
    let chartContainer = document.getElementById('chartContainer')
    let timeLabel
    // inserting all values into sensorValues variable
    let sensorValues = []
    chartData.forEach(data => {
        sensorValues.push(data.value)
    });

    //depending on what duration insert array of prefixed label
    switch (duration) {
        case "hour":
            timeLabel = [
                "60 min",
                "55 min",
                "50 min",
                "45 min",
                "40 min",
                "35 min",
                "30 min",
                "25 min",
                "20 min",
                "15 min",
                "10 min",
                "5 min",
                "0"
            ]
            break;

        case "day":

            timeLabel = [
                "24:00",
                "23:00",
                "22:00",
                "21:00",
                "20:00",
                "19:00",
                "18:00",
                "17:00",
                "16:00",
                "15:00",
                "14:00",
                "13:00",
                "12:00",
                "11:00",
                "10:00",
                "9:00",
                "8:00",
                "7:00",
                "6:00",
                "5:00",
                "4:00",
                "3:00",
                "2:00",
                "1:00",
                "0:00"
            ]
            break;

        case "week":
            timeLabel = [
                "day 7",
                "day 6",
                "day 5",
                "day 4",
                "day 3",
                "day 2",
                "day 1",
                "day 0"
            ]
            break;

        case "month":
            timeLabel = [
                "day 31",
                "day 30",
                "day 29",
                "day 28",
                "day 27",
                "day 26",
                "day 25",
                "day 24",
                "day 23",
                "day 22",
                "day 21",
                "day 20",
                "day 19",
                "day 18",
                "day 17",
                "day 16",
                "day 15",
                "day 14",
                "day 13",
                "day 12",
                "day 11",
                "day 10",
                "day 9",
                "day 8",
                "day 7",
                "day 6",
                "day 5",
                "day 4",
                "day 3",
                "day 2",
                "day 1",
                "day 0"
            ]
            break;

    }
    // creating a new chart and updating canvas element
    new Chart(chartContainer, {
        type: 'line',
        data: {
            labels: timeLabel,
            datasets: [{
                label: 'room noise level',
                data: sensorValues,
                borderWidth: 1
            }]
        },
        options: {
            responsive: true,
            scales: {
                y: {
                    beginAtZero: true
                },
            },
            plugins: {
                legend: {
                    display: false
                }
            }
        }
    });
}

export default getGraph